packages/cw3/README.md

```diff
@@ -0,0 +1,107 @@
+# CW3 Spec: MultiSig/Voting Contracts
+
+CW3 is a specification for voting contracts based on CosmWasm.
+It is an extension of CW1 (which served as an immediate 1 of N multisig).
+In this case, no key can immediately execute, but only propose
+a set of messages for execution. The proposal, subsequent
+approvals, and signature aggregation all happen on chain.  
+
+There are at least 3 different cases we want to cover in this spec:
+
+- K of N immutible multisig. One key proposes a set of messages,
+  after K-1 others approve it, it can be executed with the
+  multisig address.
+- K of N flexible, mutable multisig. Like above, but with
+  multiple contracts. One contract stores the group, which is 
+  referenced from multiple multisig contracts (which in turn
+  implement cw3). One cw3 contracts is able to update the 
+  group content (maybe needing 67% vote). Other cw3 contracts
+  may hold tokens, staking rights, etc with various execution 
+  thresholds, all controlled by one group. (Group interface
+  and updating them will be defined in a future spec, likely cw4).
+
+This should fit in this interface (possibly with some 
+extensions for pieces, but the usage should look the 
+same externally):
+
+- Token weighted voting. People lock tokens in a contract 
+  for voting rights. There is a vote threshold to execute 
+  messages. The voting set is dynamic. This has a similar
+  "propose, approve, execute" flow, but we will need to 
+  support clear YES/NO votes and quora not just absolute 
+  thresholds.
+
+The common denominator is that they allow you to propose
+arbitrary `CosmosMsg` to a contract, and allow a series
+of votes/approvals to determine if it can be executed,
+as well as a final step to execute any approved proposal once.
+
+## Base
+
+The following interfaces must be implemented for all cw3
+contracts
+
+### Messages
+
+`Propose{title, description, msgs, expires}` - This accepts 
+`Vec<CosmosMsg>` and creates a new proposal. This will return
+an auto-generated ID in the `Data` field (and the logs) that
+can be used to reference the proposal later. You can specify
+an expiration time/height in Propose, but this may be set 
+automatically by the contract (overriding or just enforcing
+min/max/default values).
+ 
+Many implementations will want to restrict who can propose.
+Maybe only people in the voting set. Maybe there is some
+deposit to be made along with the proposal. This is not
+in the spec but left open to the implementation.
```

Comment:
Should I specify that Propose does not include a yes vote from the sender if applicable?
Specify that it does only if the proposer is in the valid voter set?
Or leave it unspecified?

I'd like to specify this, as then wallets know if they need to propose and vote, or just propose.

Reply:
I think it is clean to make propose and vote two messages. It would be nice if they could be sent in one tranaction, but then we can't rely on execution data. However, this would work with client-side generated IDs.

Reply:
I defined it as the first in the README